blockchain/checkpoints.go

```diff
@@ -70,93 +70,27 @@ func (b *BlockChain) verifyCheckpoint(height int64, hash *chainhash.Hash) bool {
-	// checkpoint if the best chain is not known yet or we haven't already
-	// previously searched.
-	checkpoints := b.checkpoints
-	numCheckpoints := len(checkpoints)
-	if b.checkpointNode == nil && b.nextCheckpoint == nil {
-		// Loop backwards through the available checkpoints to find one
-		// that is already available.
-		for i := numCheckpoints - 1; i >= 0; i-- {
-			node := b.index.LookupNode(checkpoints[i].Hash)
-			if node == nil || !b.bestChain.Contains(node) {
-				continue
-			}
-
-			// Checkpoint found.  Cache it for future lookups and
-			// set the next expected checkpoint accordingly.
-			b.checkpointNode = node
-			if i < numCheckpoints-1 {
-				b.nextCheckpoint = &checkpoints[i+1]
-			}
-			return b.checkpointNode, nil
-		}
-
-		// No known latest checkpoint.  This will only happen on blocks
-		// before the first known checkpoint.  So, set the next expected
-		// checkpoint to the first checkpoint and return the fact there
-		// is no latest known checkpoint block.
-		b.nextCheckpoint = &checkpoints[0]
-		return nil, nil
-	}
-
-	// At this point we've already searched for the latest known checkpoint,
-	// so when there is no next checkpoint, the current checkpoint lockin
-	// will always be the latest known checkpoint.
-	if b.nextCheckpoint == nil {
-		return b.checkpointNode, nil
-	}
-
-	// When there is a next checkpoint and the height of the current best
-	// chain does not exceed it, the current checkpoint lockin is still
-	// the latest known checkpoint.
-	if b.bestChain.Tip().height < b.nextCheckpoint.Height {
-		return b.checkpointNode, nil
-	}
-
-	// We've reached or exceeded the next checkpoint height.  Note that
-	// once a checkpoint lockin has been reached, forks are prevented from
-	// any blocks before the checkpoint, so we don't have to worry about the
-	// checkpoint going away out from under us due to a chain reorganize.
-
-	// Cache the latest known checkpoint for future lookups.  Note that if
-	// this lookup fails something is very wrong since the chain has already
-	// passed the checkpoint which was verified as accurate before inserting
-	// it.
-	checkpointNode := b.index.LookupNode(b.nextCheckpoint.Hash)
-	if checkpointNode == nil {
-		return nil, AssertError(fmt.Sprintf("findPreviousCheckpoint "+
-			"failed lookup of known good block node %s",
-			b.nextCheckpoint.Hash))
-	}
-	b.checkpointNode = checkpointNode
-
-	// Set the next expected checkpoint.
-	checkpointIndex := -1
-	for i := numCheckpoints - 1; i >= 0; i-- {
-		if checkpoints[i].Hash.IsEqual(b.nextCheckpoint.Hash) {
-			checkpointIndex = i
-			break
-		}
+		return
 	}
-	b.nextCheckpoint = nil
-	if checkpointIndex != -1 && checkpointIndex < numCheckpoints-1 {
-		b.nextCheckpoint = &checkpoints[checkpointIndex+1]
+
+	// Nothing to update if there is no checkpoint data for the block height or
+	// the checkpoint hash does not match.
+	checkpoint, exists := b.checkpointsByHeight[node.height]
+	if !exists || node.hash != *checkpoint.Hash {
```

Comment:
I feel like the checkpoint hash not matching here is a "serious offense", in the sense that it should ordinarily not happen (unless the caller is mistaken when passing the checkpoint info during init), but if it somehow does happen we wanna know about it.

Unless this is completely going away (as you seem to imply in the messages) I'd suggest at least logging the event that the hashes don't match.

Reply:
Mismatched checkpoints are already logged by way of `verifyCheckpoint` returning false which in turn ends up returning `ErrBadCheckpoint` and the rejection gets logged.

If you feel like an additional message would be nice to have, I would say that it would probably be better over there in `verifyCheckpoint` as the goal of this function is merely to update the most recently known checkpoint while leaving rejection semantics to `verifyCheckpoint`.

That said, I would indeed generally like to continue minimizing the reliance on trusted known good checkpoints for consensus purposes because they introduce a centralized trust factor that otherwise does not exist.

The main issue is that they really serve two purposes currently.  The first is as an initial sync optimization and the second is for forcing a specific chain along with rejection of old low work blocks.

That first purpose is perfectly fine, although there are simpler ways to handle that aspect once the processing logic is decoupled from the download logic.   However, that second purpose is what I'm not really a fan of in general since, as I mentioned, since it means it essentially puts the developers in a centralized and trusted position of deciding which chain is acceptable, instead relying entirely on of the consensus mechanisms.

Reply:
> Mismatched checkpoints are already logged by way of verifyCheckpoint returning false which in turn ends up returning ErrBadCheckpoint and the rejection gets logged.

Got it. I hadn't noticed that particular sequence of calls.